In [1]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures



## Part 1. Loading the dataset

In [36]:
# Using pandas load the dataset (load remotely, not locally)
df = pd.read_csv("science_data_large.csv")
# Output the first 15 rows of the data
print(df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(df.shape)
print(df.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
(1000, 3)
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%     

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop(columns=['Size nm^3'])
y = df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## Part 3. Perform a Linear Regression

In [18]:
# Use sklearn to train a model on the training set
linearModel = LinearRegression()
linearModel.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
datapoint = [[469, 647]] #first data point from the table
prediction = linearModel.predict(datapoint)
print(f"Prediction for sample datapoint: {prediction}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
modelScore = linearModel.score(X_test, y_test)
print(f"Model's r^2 score: {modelScore}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print(f"Coefficients: {linearModel.coef_}")
print(f"Intercept: {linearModel.intercept_}")

Prediction for sample datapoint: [664984.89630433]
Model's r^2 score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


C:\Users\nekot\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


The .score() method returns the $r^2$ for for a given model. The $r^2$ score is an indication of how well the model's predictions match the actual data, and it is on a range from 0 to 1, where a number closer to 1 indicates that the model has high predictive accuracy. 

For this model, the $r^2$ score is approximately 0.855. This indicates that the Linear Regression model explains about 85.5% of the variability in the slime's size, which suggests a relatovely strong predictive accuracy.


$h(x) = -409391.48 + 866.146 \cdot$ X1 + $1032.695 \cdot$ X2 , where X1 is the Temperature and X2 is the Mols.

## Part 4. Use Cross Validation

In [29]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(linearModel, X, y, cv=100)
print(scores)
# Report on their finding and their significance

[0.72987142 0.02332136 0.64663948 0.83710031 0.81287812 0.85643292
 0.86566207 0.82943342 0.85717261 0.85818757 0.88407446 0.21899099
 0.79353519 0.93228937 0.74819809 0.91932985 0.85291091 0.72308356
 0.64174485 0.89081781 0.70643469 0.86544004 0.8450951  0.70336182
 0.86728463 0.91760202 0.90908958 0.94150897 0.83819493 0.87273024
 0.7871197  0.88467529 0.86671862 0.8176053  0.86637483 0.9025546
 0.85450024 0.88911985 0.81337559 0.62209156 0.88511368 0.87383856
 0.93874597 0.87945152 0.86422446 0.85185362 0.80632506 0.85400314
 0.90590826 0.85225702 0.65832    0.87102599 0.71313585 0.89400345
 0.58732207 0.87049636 0.88042279 0.88289086 0.34984312 0.86334775
 0.89350964 0.884054   0.83945642 0.91647922 0.84972854 0.88619777
 0.85252309 0.87032873 0.77491956 0.91480937 0.6467444  0.957948
 0.85213029 0.67112896 0.88361835 0.83713072 0.87391655 0.67181693
 0.83559998 0.92893673 0.88096535 0.74106931 0.60975577 0.42168108
 0.84981447 0.73046608 0.38714045 0.86307705 0.92088635 0.8355042

Findings: 

As I increase the number of folds (cv hyperparameter value), there are some higher scores in the iterations. For example, when cv=5, the highest score is ~0.872. When cv=10, the highest is ~0.887, and when cv=100, the highest is ~0.966. I also noticed that the scores increase and decrease in a varying manner across the iterations. For example, when for cv=100, the range is between ~0.023 to ~0.966, indicating that the model's performance is inconsistent across the folds, and is potentially overfitting in some folds, and underfitting in others.

## Part 5. Using Polynomial Regression

In [33]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)

X_poly = poly.fit_transform(X)
X_train_poly, X_test_poly, y_train_poly, y_test_poly = train_test_split(X_poly, y, test_size=0.1, random_state=42)

polynomialRegression = LinearRegression()
polynomialRegression.fit(X_train_poly, y_train_poly)


# Report on the metrics and output the resultant equation as you did in Part 3.
print(f"Feature names: {poly.get_feature_names_out()}\n")

print(f"Model's score: {polynomialRegression.score(X_test_poly, y_test_poly)}\n")

print(f"Coefficients: {polynomialRegression.coef_}\n")
print(f"Intercept: {polynomialRegression.intercept_}")


Feature names: ['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']

Model's score: 1.0

Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27197088e-07  1.26458843e-11
  2.00000000e+00  2.85714287e-02]

Intercept: 2.0479375962167978e-05


The model's $r^2$ score is 1.0, indicating that it explains 100% of the variability in the data, indicating that it has a 100% predictive accuracy. While this may seem optimistic, it is important to also consider that the model might be overfitting to the training data, and may not generalize and perform well on new, unseen data. 

$h(x) = 12$ $\cdot$ X1 $+$ $2$ $\cdot$ X4 $+$ $0.0286$ $\cdot$ X5, where X1 = Temperature, X4 = Temperature $\cdot$ Mols, X5 = Mols^2